File: src/rvgen/utils/elfbuilder.py
# Refer to https://en.wikipedia.org/wiki/Executable_and_Linkable_Format

from dataclasses import dataclass
from pathlib import Path
import struct

# Section flags
SHF_WRITE     = 0x1
SHF_ALLOC     = 0x2
SHF_EXECINSTR = 0x4

# Segment types & flags
PT_LOAD = 1
PF_X    = 1
PF_W    = 2
PF_R    = 4

# [Constants] Layout constants for ELF32-littleriscv
_E32_EHDR_SIZE = 52
_E32_PHDR_SIZE = 32
_E32_SHDR_SIZE = 40

# [Constants] Layout constants for ELF64-littleriscv
_E64_EHDR_SIZE = 64
_E64_PHDR_SIZE = 56
_E64_SHDR_SIZE = 64


@dataclass
class ElfSection:
    name: str
    inbytes: bytes
    addr: int = 0x0
    flags: int = SHF_ALLOC | SHF_EXECINSTR  # SHF_ALLOC | SHF_EXECINSTR
    align: int = 4
    section_type: int = 1
    link: int = 0
    info: int = 0
    entsize: int = 0


@dataclass
class ElfSymbol:
    """Global symbol defined at an offset within a zero-based section index."""
    name: str
    section: int
    offset: int = 0
    size: int = 0
    symbol_type: int = 1
    

class ElfBuilder:
    def __init__(self) -> None:
        pass

    def build_with_symbols(self, sections: list[ElfSection], symbols: list[ElfSymbol],
                           is_64bit: bool = True, start_addr: int = 0x80000000) -> bytes:
        sections = list(sections)
        if any(section.name in {".strtab", ".symtab"} for section in sections):
            raise ValueError("symbol table section names are reserved")
        strings = bytearray(b"\0")
        entry_size = 24 if is_64bit else 16
        table = bytearray(entry_size)
        names = set()
        for symbol in symbols:
            if (not symbol.name or not symbol.name.isascii() or "\0" in symbol.name
                    or symbol.name in names or not 0 <= symbol.symbol_type <= 15):
                raise ValueError("invalid or duplicate ELF symbol")
            names.add(symbol.name)
            if not 0 <= symbol.section < len(sections):
                raise ValueError("invalid symbol section")
            section = sections[symbol.section]
            if min(symbol.offset, symbol.size) < 0 or symbol.offset + symbol.size > len(section.inbytes):
                raise ValueError("symbol lies outside its section")
            name_offset = len(strings)
            strings.extend(symbol.name.encode("ascii") + b"\0")
            value = section.addr + symbol.offset
            if is_64bit:
                table.extend(struct.pack('<IBBHQQ', name_offset, 0x10 | symbol.symbol_type,
                                         0, symbol.section + 1, value, symbol.size))
            else:
                table.extend(struct.pack('<IIIBBH', name_offset, value, symbol.size,
                                         0x10 | symbol.symbol_type, 0, symbol.section + 1))
        string_index = len(sections) + 1
        sections.append(ElfSection('.strtab', bytes(strings), flags=0, align=1, section_type=3))
        sections.append(ElfSection('.symtab', bytes(table), flags=0, align=8 if is_64bit else 4,
                                   section_type=2, link=string_index, info=1, entsize=entry_size))
        return self.build(sections, is_64bit, start_addr)

    @staticmethod
    def _sh_flags_to_p_flags(sh_flags: int) -> int:
        p_flags = PF_R
        if sh_flags & SHF_WRITE:
            p_flags |= PF_W
        if sh_flags & SHF_EXECINSTR:
            p_flags |= PF_X
        return p_flags

    def build(self, sections: list[ElfSection], is_64bit: bool = True, start_addr: int = 0x80000000) -> bytes:
        """
        This function builds a little-endian RISCV executable ELF file.

        The layout of the ELF file is as follows:
        - ELF header
        - Program headers
        - Section data
        - .shstrtab
        - Section headers
        """
        # 1. Build .shstrtab dynamically
        shstrtab = bytearray(b"\x00")
        name_offsets = {}
        for s in sections:
            name_offsets[s.name] = len(shstrtab)
            shstrtab += s.name.encode("ascii") + b"\x00"
        name_offsets[".shstrtab"] = len(shstrtab)
        shstrtab.extend(b".shstrtab\x00")

        # Structure sizes
        ehdr_size = _E64_EHDR_SIZE if is_64bit else _E32_EHDR_SIZE
        phdr_size = _E64_PHDR_SIZE if is_64bit else _E32_PHDR_SIZE
        shdr_size = _E64_SHDR_SIZE if is_64bit else _E32_SHDR_SIZE

        phoff = ehdr_size
        num_phdrs = len([s for s in sections if s.flags & SHF_ALLOC])

        cur_offset = phoff + num_phdrs * phdr_size
        
        # 2. Compute file offsets for each section
        section_offsets = []
        for s in sections:
            if s.align > 1 and (cur_offset % s.align) != 0:
                cur_offset += s.align - (cur_offset % s.align)
            section_offsets.append(cur_offset)
            cur_offset += len(s.inbytes)

        # Place .shstrtab after all other sections
        shstrtab_offset = cur_offset
        cur_offset += len(shstrtab)

        if (cur_offset % 8) != 0:
            cur_offset += 8 - (cur_offset % 8)
        shoff = cur_offset
        num_shdrs = len(sections) + 2  # + NULL, .shstrtab

        # 3. Pack program headers
        phdrs_bytes = bytearray()
        for s, offset in zip(sections, section_offsets):
            if not (s.flags & SHF_ALLOC):
                continue
            p_flags = self._sh_flags_to_p_flags(s.flags)
            size = len(s.inbytes)
            if is_64bit:
                phdrs_bytes.extend(struct.pack(
                    '<IIQQQQQQ',
                    PT_LOAD, p_flags, offset, s.addr, s.addr, size, size, s.align
                ))
            else:
                phdrs_bytes.extend(struct.pack(
                    '<IIIIIIII',
                    PT_LOAD, offset, s.addr, s.addr, size, size, p_flags, s.align
                ))

        # 4. Pack section headers
        shdrs_bytes = bytearray()
        shdrs_bytes.extend(b"\x00" * shdr_size)

        for s, offset in zip(sections, section_offsets):
            sh_name = name_offsets[s.name]
            sh_type = s.section_type
            size = len(s.inbytes)
            if is_64bit:
                shdrs_bytes.extend(struct.pack(
                    '<IIQQQQIIQQ',
                    sh_name, sh_type, s.flags, s.addr, offset, size, s.link, s.info, s.align, s.entsize
                ))
            else:
                shdrs_bytes.extend(struct.pack(
                    '<IIIIIIIIII',
                    sh_name, sh_type, s.flags, s.addr, offset, size, s.link, s.info, s.align, s.entsize
                ))

        # .shstrtab section header (index = len(sections) + 1)
        shstrtab_idx = len(sections) + 1
        if is_64bit:
            shdrs_bytes.extend(struct.pack(
                '<IIQQQQIIQQ',
                name_offsets[".shstrtab"], 3, 0, 0, shstrtab_offset, len(shstrtab), 0, 0, 1, 0
            ))
        else:
            shdrs_bytes.extend(struct.pack(
                '<IIIIIIIIII',
                name_offsets[".shstrtab"], 3, 0, 0, shstrtab_offset, len(shstrtab), 0, 0, 1, 0
            ))

        # 5. Pack the ELF header
        if is_64bit:
            e_ident = b"\x7fELF\x02\x01\x01\x00" + b"\x00" * 8
            ehdr_bytes = e_ident + struct.pack(
                '<HHIQQQIHHHHHH',
                2, 0xf3, 1, start_addr, phoff, shoff, 0,
                ehdr_size, phdr_size, num_phdrs, shdr_size, num_shdrs, shstrtab_idx
            )
        else:
            e_ident = b"\x7fELF\x01\x01\x01\x00" + b"\x00" * 8
            ehdr_bytes = e_ident + struct.pack(
                '<HHIIIIIHHHHHH',
                2, 0xf3, 1, start_addr, phoff, shoff, 0,
                ehdr_size, phdr_size, num_phdrs, shdr_size, num_shdrs, shstrtab_idx
            )

        # 6. Concatenate everything
        file_bytes = bytearray(ehdr_bytes + phdrs_bytes)
        for s, offset in zip(sections, section_offsets):
            if len(file_bytes) < offset:
                file_bytes.extend(b"\x00" * (offset - len(file_bytes)))
            file_bytes.extend(s.inbytes)

        if len(file_bytes) < shstrtab_offset:
            file_bytes.extend(b"\x00" * (shstrtab_offset - len(file_bytes)))
        file_bytes.extend(shstrtab)

        if len(file_bytes) < shoff:
            file_bytes.extend(b"\x00" * (shoff - len(file_bytes)))
        file_bytes.extend(shdrs_bytes)
        
        return bytes(file_bytes)

    def save(self, elf_bytes: bytes, destination_path: str | Path) -> None:
        with open(destination_path, 'wb') as f:
            f.write(elf_bytes)


# def gen_elf(inbytes: bytes, start_addr: int, section_addr: int, destination_path: str, is_64bit: bool = True) -> None:
#     """
#     Generate a RISCV executable ELF file.

#     Args:
#         inbytes: The bytes to put into the ELF file.
#                  Must be in little endian format.
#         start_addr: The entry PC.
#         section_addr: The address at which to load the .text.init section.
#                       May be None to use ``start_addr``.
#         destination_path: Output ELF path.
#         is_64bit: If True emit ELF64-littleriscv, else ELF32-littleriscv.
#     """
#     if section_addr is None:
#         section_addr = start_addr

#     if is_64bit:
#         elf_bytes = _build_elf64_riscv(inbytes, start_addr, section_addr)
#     else:
#         elf_bytes = _build_elf32_riscv(inbytes, start_addr, section_addr)

#     with open(destination_path, 'wb') as f:
#         f.write(elf_bytes)

File: rvgen/src/utils/elfbuilder.rs
//! Little-endian RISC-V ELF32/ELF64 executable writer.
//! Layout and header defaults match the original Python ElfBuilder.

use std::{collections::HashSet, path::Path};

use crate::Result;

pub const SHF_WRITE: u64 = 1;
pub const SHF_ALLOC: u64 = 2;
pub const SHF_EXECINSTR: u64 = 4;
pub const PT_LOAD: u32 = 1;
pub const PF_X: u32 = 1;
pub const PF_W: u32 = 2;
pub const PF_R: u32 = 4;

#[derive(Debug, Clone)]
pub struct ElfSection {
    pub name: String,
    pub inbytes: Vec<u8>,
    pub addr: u64,
    pub flags: u64,
    pub align: u64,
    pub section_type: u32,
    pub link: u32,
    pub info: u32,
    pub entsize: u64,
}

impl ElfSection {
    pub fn new(name: impl Into<String>, inbytes: impl Into<Vec<u8>>) -> Self {
        Self {
            name: name.into(),
            inbytes: inbytes.into(),
            addr: 0,
            flags: SHF_ALLOC | SHF_EXECINSTR,
            align: 4,
            section_type: 1,
            link: 0,
            info: 0,
            entsize: 0,
        }
    }
}

/// A global symbol defined at an offset within a section (zero-based index).
#[derive(Debug, Clone)]
pub struct ElfSymbol {
    pub name: String,
    pub section: usize,
    pub offset: u64,
    pub size: u64,
    /// ELF STT_OBJECT = 1, STT_FUNC = 2.
    pub symbol_type: u8,
}

#[derive(Debug, Default)]
pub struct ElfBuilder;

fn add(a: u64, b: u64) -> Result<u64> {
    a.checked_add(b)
        .ok_or_else(|| "ELF layout exceeds 64 bits".into())
}

fn align_up(offset: u64, alignment: u64) -> Result<u64> {
    let alignment = alignment.max(1);
    add(offset, (alignment - offset % alignment) % alignment)
}

fn w16(bytes: &mut Vec<u8>, value: u16) {
    bytes.extend_from_slice(&value.to_le_bytes());
}

fn w32(bytes: &mut Vec<u8>, value: u32) {
    bytes.extend_from_slice(&value.to_le_bytes());
}

fn w64(bytes: &mut Vec<u8>, value: u64) {
    bytes.extend_from_slice(&value.to_le_bytes());
}

fn word(bytes: &mut Vec<u8>, value: u64, is_64bit: bool) -> Result<()> {
    if is_64bit {
        w64(bytes, value);
    } else {
        w32(
            bytes,
            value.try_into().map_err(|_| "value does not fit ELF32")?,
        );
    }
    Ok(())
}

impl ElfBuilder {
    pub fn build_with_symbols(
        &self,
        sections: &[ElfSection],
        symbols: &[ElfSymbol],
        is_64bit: bool,
        start_addr: u64,
    ) -> Result<Vec<u8>> {
        let mut sections = sections.to_vec();
        let mut names = HashSet::new();
        let mut strings = vec![0];
        let entry_size = if is_64bit { 24 } else { 16 };
        let mut table = vec![0; entry_size]; // STN_UNDEF, the only local symbol
        for symbol in symbols {
            if symbol.name.is_empty()
                || !symbol.name.is_ascii()
                || symbol.name.contains('\0')
                || !names.insert(&symbol.name)
                || symbol.symbol_type > 15
            {
                return Err("invalid or duplicate ELF symbol".into());
            }
            let section = sections
                .get(symbol.section)
                .ok_or("invalid symbol section")?;
            if add(symbol.offset, symbol.size)? > section.inbytes.len() as u64 {
                return Err("symbol lies outside its section".into());
            }
            let value = add(section.addr, symbol.offset)?;
            let index = u16::try_from(symbol.section + 1)?;
            w32(&mut table, u32::try_from(strings.len())?);
            strings.extend_from_slice(symbol.name.as_bytes());
            strings.push(0);
            if is_64bit {
                table.extend_from_slice(&[0x10 | symbol.symbol_type, 0]);
                w16(&mut table, index);
                w64(&mut table, value);
                w64(&mut table, symbol.size);
            } else {
                word(&mut table, value, false)?;
                word(&mut table, symbol.size, false)?;
                table.extend_from_slice(&[0x10 | symbol.symbol_type, 0]);
                w16(&mut table, index);
            }
        }
        let string_index = u32::try_from(sections.len() + 1)?;
        let mut strtab = ElfSection::new(".strtab", strings);
        strtab.flags = 0;
        strtab.align = 1;
        strtab.section_type = 3;
        sections.push(strtab);
        let mut symtab = ElfSection::new(".symtab", table);
        symtab.flags = 0;
        symtab.align = if is_64bit { 8 } else { 4 };
        symtab.section_type = 2;
        symtab.link = string_index;
        symtab.info = 1;
        symtab.entsize = entry_size as u64;
        sections.push(symtab);
        self.build(&sections, is_64bit, start_addr)
    }

    pub fn sh_flags_to_p_flags(flags: u64) -> u32 {
        PF_R | if flags & SHF_WRITE != 0 { PF_W } else { 0 }
            | if flags & SHF_EXECINSTR != 0 { PF_X } else { 0 }
    }

    pub fn build(
        &self,
        sections: &[ElfSection],
        is_64bit: bool,
        start_addr: u64,
    ) -> Result<Vec<u8>> {
        // Extended section numbering is intentionally outside this small writer's scope.
        if sections.len() + 2 >= 0xff00 {
            return Err("too many ELF sections".into());
        }
        let (ehsize, phsize, shsize) = if is_64bit { (64, 56, 64) } else { (52, 32, 40) };
        let phnum = sections.iter().filter(|s| s.flags & SHF_ALLOC != 0).count();
        let mut names = HashSet::new();
        let mut shstrtab = vec![0];
        let mut name_offsets = Vec::new();
        let mut section_offsets = Vec::new();
        let mut offset = ehsize + phnum as u64 * phsize;
        for section in sections {
            if section.name.is_empty()
                || !section.name.is_ascii()
                || section.name.contains('\0')
                || section.name == ".shstrtab"
                || !names.insert(&section.name)
            {
                return Err(
                    format!("invalid or duplicate section name: {:?}", section.name).into(),
                );
            }
            if section.align > 1 && !section.align.is_power_of_two() {
                return Err(
                    format!("section {} alignment must be a power of two", section.name).into(),
                );
            }
            let end_addr = add(section.addr, section.inbytes.len() as u64)?;
            if !is_64bit && (section.addr > u32::MAX as u64 || end_addr > 1_u64 << 32) {
                return Err("section address does not fit ELF32".into());
            }
            name_offsets.push(u32::try_from(shstrtab.len())?);
            shstrtab.extend_from_slice(section.name.as_bytes());
            shstrtab.push(0);
            offset = align_up(offset, section.align)?;
            section_offsets.push(offset);
            offset = add(offset, section.inbytes.len() as u64)?;
        }
        let shstrtab_name = u32::try_from(shstrtab.len())?;
        shstrtab.extend_from_slice(b".shstrtab\0");
        let shstrtab_offset = offset;
        let shoff = align_up(add(offset, shstrtab.len() as u64)?, 8)?;
        let file_size = add(shoff, (sections.len() as u64 + 2) * shsize)?;
        if !is_64bit && (file_size > u32::MAX as u64 || start_addr > u32::MAX as u64) {
            return Err("ELF layout or entry address does not fit ELF32".into());
        }
        let file_size = usize::try_from(file_size)?;
        let mut bytes = Vec::new();
        bytes.try_reserve_exact(file_size)?;

        bytes.extend_from_slice(b"\x7fELF");
        bytes.extend_from_slice(&[if is_64bit { 2 } else { 1 }, 1, 1, 0]);
        bytes.extend_from_slice(&[0; 8]);
        w16(&mut bytes, 2); // ET_EXEC
        w16(&mut bytes, 0xf3); // EM_RISCV
        w32(&mut bytes, 1);
        word(&mut bytes, start_addr, is_64bit)?;
        word(&mut bytes, ehsize, is_64bit)?;
        word(&mut bytes, shoff, is_64bit)?;
        w32(&mut bytes, 0); // e_flags
        for value in [
            ehsize as u16,
            phsize as u16,
            phnum as u16,
            shsize as u16,
            (sections.len() + 2) as u16,
            (sections.len() + 1) as u16,
        ] {
            w16(&mut bytes, value);
        }

        for (section, &file_offset) in sections.iter().zip(&section_offsets) {
            if section.flags & SHF_ALLOC == 0 {
                continue;
            }
            let flags = Self::sh_flags_to_p_flags(section.flags);
            w32(&mut bytes, PT_LOAD);
            if is_64bit {
                w32(&mut bytes, flags);
            }
            for value in [
                file_offset,
                section.addr,
                section.addr,
                section.inbytes.len() as u64,
                section.inbytes.len() as u64,
            ] {
                word(&mut bytes, value, is_64bit)?;
            }
            if !is_64bit {
                w32(&mut bytes, flags);
            }
            word(&mut bytes, section.align, is_64bit)?;
        }
        for (section, &file_offset) in sections.iter().zip(&section_offsets) {
            bytes.resize(file_offset as usize, 0);
            bytes.extend_from_slice(&section.inbytes);
        }
        bytes.extend_from_slice(&shstrtab);
        bytes.resize((shoff + shsize) as usize, 0); // alignment and null section header

        let string_section = ElfSection {
            name: ".shstrtab".into(),
            inbytes: shstrtab,
            addr: 0,
            flags: 0,
            align: 1,
            section_type: 3,
            link: 0,
            info: 0,
            entsize: 0,
        };
        for ((section, name_offset), file_offset) in sections
            .iter()
            .chain([&string_section])
            .zip(name_offsets.into_iter().chain([shstrtab_name]))
            .zip(section_offsets.into_iter().chain([shstrtab_offset]))
        {
            w32(&mut bytes, name_offset);
            w32(&mut bytes, section.section_type);
            for value in [
                section.flags,
                section.addr,
                file_offset,
                section.inbytes.len() as u64,
            ] {
                word(&mut bytes, value, is_64bit)?;
            }
            w32(&mut bytes, section.link);
            w32(&mut bytes, section.info);
            word(&mut bytes, section.align, is_64bit)?;
            word(&mut bytes, section.entsize, is_64bit)?;
        }
        debug_assert_eq!(bytes.len(), file_size);
        Ok(bytes)
    }

    pub fn save(&self, elf_bytes: &[u8], destination_path: impl AsRef<Path>) -> Result<()> {
        let path = destination_path.as_ref();
        std::fs::write(path, elf_bytes)
            .map_err(|error| format!("cannot write {}: {error}", path.display()).into())
    }
}
